src/board.ts

```diff
@@ -334,6 +335,155 @@ export function getKeyAtDomPos(pos: cg.NumberPair, asWhite: boolean, bounds: Cli
+  if (targetKey === undefined) return false;
+  const pos = key2pos(targetKey);
+  // +
+  if (orig[0] === pos[0] || orig[1] === pos[1]) return true;
+  // x
+  if (Math.abs(orig[0] - pos[0]) === Math.abs(orig[1] - pos[1])) return true;
+  // knight
+  if (Math.abs(orig[0] - pos[0]) === 2 && Math.abs(orig[1] - pos[1]) === 1) return true;
+  if (Math.abs(orig[0] - pos[0]) === 1 && Math.abs(orig[1] - pos[1]) === 2) return true;
+  // all other
+  return false;
+}
+
+export function getSnappedKeyAtDomPos(orig: cg.Key, pos: cg.NumberPair, asWhite: boolean, bounds: ClientRect): cg.Key | undefined {
+  // 1. Get key at dom pos
+  // 2. If move is valid, short circuit
+  // 3. If move is not valid, pick nearest angle and snap distance to dragged distance
+  // Only snapping to + and x currently.
+  // TODO: also snap to knight angles, but stop after going past knight tile?
+  // TODO: cleanly roll these if statements into 1 loop?
+
+  const unsnappedKey = getKeyAtDomPos(pos, asWhite, bounds);
+  const origRowCol = key2pos(orig);
+  if (isAlreadySnapped(origRowCol, unsnappedKey)) return unsnappedKey;
+
+  const draggedOutOfBox = unsnappedKey === undefined;
+  const unsnappedRowCol = unsnappedKey === undefined ? origRowCol : key2pos(unsnappedKey);
+  const origCenterCoord = computeSquareCenter(orig, asWhite, bounds);
+
+  // Use degrees for ease of round numbers in code
+  const origCenterToMouseDegrees = Math.atan2(pos[1] - origCenterCoord[1], origCenterCoord[0] - pos[0]) * 180 / Math.PI + 180;
+  const origCenterToMouseCoordX = origCenterCoord[0] - pos[0];
+  const origCenterToMouseCoordY = origCenterCoord[1] - pos[1];
+
+  const origCenterToMouseCoordDistance = Math.sqrt(
+      origCenterToMouseCoordX * origCenterToMouseCoordX +
+      origCenterToMouseCoordY * origCenterToMouseCoordY
+  );
+  const diagonalCoordDistance = Math.sqrt(
+      bounds.width * bounds.width +
+      bounds.height * bounds.height
+  ) / 8;
+  const squareCoordDistance = Math.floor((origCenterToMouseCoordDistance + diagonalCoordDistance / 2) / diagonalCoordDistance);
+
+  if (origCenterToMouseDegrees < (45 * 0 + 45 / 2)) {
+    // console.log('e+');
+    if (draggedOutOfBox) {
+      return pos2key([asWhite ? 7 : 0, origRowCol[1]]);
+    }
+    return pos2key([unsnappedRowCol[0], origRowCol[1]]);
+  }
+  if (origCenterToMouseDegrees < (45 * 1 + 45 / 2)) {
+    // console.log('ne');
+    if (asWhite) {
+      let newPos: cg.NumberPair = [origRowCol[0] + squareCoordDistance, origRowCol[1] + squareCoordDistance];
+      while (newPos[0] > 7 || newPos[1] > 7) {
+        newPos = [newPos[0] - 1, newPos[1] - 1];
+      }
+      return pos2key(newPos);
+    } else {
+      let newPos: cg.NumberPair = [origRowCol[0] - squareCoordDistance, origRowCol[1] - squareCoordDistance];
+      while (newPos[0] < 0 || newPos[1] < 0) {
+        newPos = [newPos[0] + 1, newPos[1] + 1];
+      }
+      return pos2key(newPos);
+    }
+  }
+  if (origCenterToMouseDegrees < (45 * 2 + 45 / 2)) {
```

Comment:
BTW these literal expressions get compiled down to a single value. I wrote the numbers out this way in case it helps roll the if statements into a loop